### DISCUSSION TITLE
How to use derived/computed state?

### DISCUSSION THREAD

**Discussion by vlad-solomon** [pmndrs/zustand#1694] [Q&A] [answered]
I'm fairly new to Zustand and really loving it so far, but I've come to a grinding halt when I try to figure out how to derive state. I've set up a really simple example here: https://codesandbox.io/s/cool-volhard-z2c1jh?file=/src/App.js

This is the store
```js
export const useStore = create((set) => ({
  people: [
    {
      name: "James",
      age: 23
    },
    {
      name: "Paul",
      age: 18
    },
    {
      name: "Brian",
      age: 50
    },
    {
      name: "Aaron",
      age: 20
    }
  ]
}));
```

What I'm trying to achieve is to have the store map over the ```people``` and display that in two lists, one unordered and one ordered by age. The problem is that when I try to sort the ordered array, it sorts the state as a whole and turns both lists into ordered ones.

What I want to achieve is this:
![image](https://user-images.githubusercontent.com/48878155/225296934-8924854f-4793-4f4f-90c1-43bec76aa204.png)

But, like I said, when I sort the array I get this (the state itself sorts and modifies both):
![image](https://user-images.githubusercontent.com/48878155/225297085-a6c7fc4f-0890-4af3-b573-7929cbb2f79a.png)

If anybody can help with this problem I will be very grateful. Thank you 😊

**Comment by dai-shi** [comment, marked as answer]
`sort` in JavaScript modifies array in place. https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/sort
You want to copy it in advance.
```js
const sortedArray = [...array].sort(...)
```

**Reply by vlad-solomon** [reply]
Oh wow, I can't figure out how I missed that. I think I wrote ```[...array.sort(...)]``` instead of ```[...array].sort(...)```. You've been a great help, thank you very much! 🥰🥰

**Reply by crswll** [reply]
You can also use `array.toSorted()`.
